File: libs/parsing/src/parsing/stitch.py
"""Stitch shard JSONs back into one DoclingDocument (PRD §6.4 step 1).

Heading hierarchy must be reconciled across boundaries or every chunk
after shard 1 loses its section context. Docling serialises each shard
with its own heading tree restart, so we rebuild the document by
concatenating shard texts in order and dropping the duplicate leading
heading when a shard starts on the same heading the previous one ended
on (the 1-page overlap shows up here too).
"""

from __future__ import annotations

import json
from typing import Any


def load_shard_docs(
    fetch: dict[int, str],
) -> list[dict[str, Any]]:
    """Parse shard JSON payloads ``{idx: json_text}`` in idx order.

    Malformed shard JSON raises here — the embed job must not silently
    drop a shard's content; the retry ladder handles it upstream.
    """
    docs = []
    for idx in sorted(fetch):
        text = fetch[idx]
        try:
            docs.append(json.loads(text))
        except json.JSONDecodeError:
            # Parser uploads export_to_markdown() text directly — treat a
            # non-JSON payload as the markdown body itself.
            docs.append({"markdown": text})
    return docs


def stitch(
    docs: list[dict[str, Any]],
    shard_page_ranges: list[tuple[int, int]] | None = None,
) -> dict[str, Any]:
    """Concatenate shard documents in order into a single doc-like dict.

    Works on Docling's serialised export shape ({"texts": [...], ...}),
    tolerating shards that produced no text blocks. Heading reconciliation
    is line-based: a shard's first markdown heading that equals the
    previous shard's last is dropped (overlap dedupe at the boundary).

    ``shard_page_ranges`` (optional) is a list of ``(page_start, page_end)``
    1-based inclusive pairs aligned with ``docs`` (doc order = sorted shard
    idx). When given, the result carries ``"pages"``: a list mapping the
    0-based line index of the stitched markdown to a 1-based page number,
    interpolated proportionally within each shard. Ranges align to the
    docs that actually contributed lines — a shard with empty markdown is
    skipped together with its range so the map stays monotone.
    """
    if not docs:
        return {"texts": [], "markdown": ""}

    lines: list[str] = []
    page_of_line: list[int] = []
    # Pair each doc with its range in lockstep: a shard with empty markdown
    # contributes neither lines nor a consumed range, so the map stays
    # monotone even when some shards produced no text.
    paired = (
        zip(docs, shard_page_ranges, strict=True) if shard_page_ranges is not None else
        ((doc, None) for doc in docs)
    )
    for doc, page_range in paired:
        md = _markdown_of(doc)
        if not md:
            continue
        doc_lines = md.splitlines()
        if lines and doc_lines:
            # drop a duplicated boundary heading from the overlap
            prev_last = _last_heading(lines)
            first = _first_heading(doc_lines)
            if prev_last is not None and first == prev_last:
                doc_lines = doc_lines[1:]
                # and the blank line under it, if any
                while doc_lines and not doc_lines[0].strip():
                    doc_lines = doc_lines[1:]
        lines.extend(doc_lines)
        if page_range is not None:
            page_start, page_end = page_range
            n = len(doc_lines)
            span = page_end - page_start + 1
            for local in range(n):
                page_of_line.append(page_start + (local * span) // n)

    out: dict[str, Any] = {"texts": [], "markdown": "\n".join(lines).strip() + "\n"}
    if shard_page_ranges is not None:
        out["pages"] = page_of_line
    return out


def _markdown_of(doc: dict[str, Any]) -> str:
    md = doc.get("markdown")
    if isinstance(md, str):
        return md
    # Fall back to DoclingDocument export_text/markdown nested shapes.
    return ""


def _first_heading(lines: list[str]) -> str | None:
    for line in lines:
        s = line.strip()
        if s.startswith("#"):
            return s
    return None


def _last_heading(lines: list[str]) -> str | None:
    for line in reversed(lines):
        s = line.strip()
        if s.startswith("#"):
            return s
    return None

File: internal/pipeline/stitch.go
package pipeline

import (
	"encoding/json"
	"sort"
	"strings"
)

// StitchedDoc is the stitch output: one markdown string plus a per-line
// page map (0-based line index → 1-based page).
type StitchedDoc struct {
	Markdown string
	// Pages is aligned with Markdown's lines: Pages[i] is the 1-based page of
	// line i, where the single trailing "\n" that Markdown always ends with
	// does not start a new line. So len(Pages) == number of "\n" in Markdown
	// == len(strings.Split(strings.TrimSuffix(Markdown, "\n"), "\n")).
	//
	// It is nil when no shard contributed a page range. Consumers that index it
	// must treat a nil or short map as "no page information" rather than
	// guessing — the chunker's pageMapper does exactly that.
	Pages []int
}

// Stitch concatenates shard markdown payloads in order into one doc.
//
// Heading reconciliation is line-based: a shard's first markdown heading
// that equals the previous shard's last is dropped (the 1-page overlap
// shows up here). shardPageRanges is a list of (page_start, page_end)
// 1-based inclusive pairs aligned with docs order (sorted shard idx). The
// result carries the per-line page map interpolated proportionally within
// each shard; ranges align to the docs that actually contributed lines — a
// shard with empty markdown is skipped together with its range so the map
// stays monotone. Port of stitch.py.
func Stitch(docs map[int]string, shardPageRanges map[int][2]int) StitchedDoc {
	idxs := make([]int, 0, len(docs))
	for idx := range docs {
		idxs = append(idxs, idx)
	}
	sort.Ints(idxs)

	var lines []string
	var pageOfLine []int

	for _, idx := range idxs {
		md := normalizeMarkdown(docs[idx])
		if md == "" {
			continue
		}
		// A shard payload almost always ends in "\n" (docling's markdown does).
		// strings.Split would then produce a trailing empty element that is not
		// a real line — emitting a page entry for it inflates the map by one per
		// shard and shifts every subsequent page number. Normalize it away
		// before splitting so len(docLines) is the true line count.
		md = strings.TrimSuffix(md, "\n")
		docLines := strings.Split(md, "\n")
		if len(lines) > 0 && len(docLines) > 0 {
			// drop a duplicated boundary heading from the overlap
			prevLast := lastHeading(lines)
			first := firstHeading(docLines)
			if prevLast != nil && first != nil && *first == *prevLast {
				docLines = docLines[1:]
				// and the blank line under it, if any
				for len(docLines) > 0 && strings.TrimSpace(docLines[0]) == "" {
					docLines = docLines[1:]
				}
			}
		}
		lines = append(lines, docLines...)
		if pr, ok := shardPageRanges[idx]; ok {
			pageStart, pageEnd := pr[0], pr[1]
			n := len(docLines)
			if n == 0 {
				continue
			}
			span := pageEnd - pageStart + 1
			for local := 0; local < n; local++ {
				pageOfLine = append(pageOfLine, pageStart+(local*span)/n)
			}
		}
	}
	md := strings.TrimSpace(strings.Join(lines, "\n")) + "\n"
	// TrimSpace can drop leading/trailing blank lines from the join, which would
	// leave Pages indexed against lines that are no longer in Markdown. Shift the
	// map to match, or drop it when the two cannot be reconciled.
	pageOfLine = alignPages(md, pageOfLine, lines)
	return StitchedDoc{Markdown: md, Pages: pageOfLine}
}

// alignPages reconciles a page map built from the pre-trim line list with the
// final markdown.
//
// Markdown is the joined lines with surrounding whitespace trimmed plus one
// trailing "\n", so the only possible discrepancy is dropped leading or trailing
// blank lines; the map is shifted by that amount. When the sizes still cannot be
// reconciled the map is dropped entirely — a missing page is better than a wrong
// one, and every consumer treats nil as "no page information".
func alignPages(markdown string, pageOfLine []int, lines []string) []int {
	if len(pageOfLine) == 0 {
		return nil
	}
	emitted := len(lines)
	// Count the leading lines TrimSpace removed from the join.
	lead := 0
	for lead < len(lines) && strings.TrimSpace(lines[lead]) == "" {
		lead++
	}
	if lead > 0 && lead <= len(pageOfLine) {
		pageOfLine = pageOfLine[lead:]
		emitted -= lead
	}
	// Trailing blank lines are trimmed too.
	trail := 0
	for trail < len(lines)-lead && strings.TrimSpace(lines[len(lines)-1-trail]) == "" {
		trail++
	}
	if trail > 0 && trail <= len(pageOfLine) {
		pageOfLine = pageOfLine[:len(pageOfLine)-trail]
		emitted -= trail
	}
	if len(pageOfLine) != emitted || len(pageOfLine) != strings.Count(markdown, "\n") {
		return nil
	}
	return pageOfLine
}

// normalizeMarkdown tolerates the two payload shapes the parser uploads:
// plain markdown, or a JSON wrapper {"markdown": ...} (load_shard_docs
// treated non-JSON payloads as the markdown body itself; JSON payloads got
// their "markdown" field).
func normalizeMarkdown(text string) string {
	trimmed := strings.TrimSpace(text)
	if !strings.HasPrefix(trimmed, "{") {
		return text
	}
	var payload map[string]any
	if err := json.Unmarshal([]byte(trimmed), &payload); err != nil {
		return text
	}
	if md, ok := payload["markdown"].(string); ok {
		return md
	}
	return text
}

func firstHeading(lines []string) *string {
	for _, line := range lines {
		s := strings.TrimSpace(line)
		if strings.HasPrefix(s, "#") {
			return &s
		}
	}
	return nil
}

func lastHeading(lines []string) *string {
	for i := len(lines) - 1; i >= 0; i-- {
		s := strings.TrimSpace(lines[i])
		if strings.HasPrefix(s, "#") {
			return &s
		}
	}
	return nil
}
